Implementation of the Morphological Snakes

pmneila opened the pull request:
## Description

Implementation of the Morphological Snakes for image segmentation as suggested by @soupault in pmneila/morphsnakes#10. Details about the algorithm and some animated examples can be found in the [original morphsnakes repo](https://github.com/pmneila/morphsnakes).

![sphx_glr_plot_morphsnakes_001](https://user-images.githubusercontent.com/1254337/30060179-87580986-9242-11e7-8877-6542a9ca8fbe.png)

## Checklist

- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

## References

This is a new feature implementing the algorithms from the paper:

> _A Morphological Approach to Curvature-based Evolution of Curves and Surfaces_, Pablo Márquez-Neila, Luis Baumela and Luis Álvarez. In IEEE Transactions on Pattern Analysis and Machine Intelligence (TPAMI), Vol. 36, No. 1. (January 2014), pp. 2-17.

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @pmneila! Thanks for submitting the PR.





 - In the file [`skimage/segmentation/morphsnakes.py`](https://github.com/scikit-image/scikit-image/blob/4adaadc582c7a74454d5cb654a8dfaf605b81393/skimage/segmentation/morphsnakes.py), following are the PEP8 issues :

> [Line 155:25](https://github.com/scikit-image/scikit-image/blob/4adaadc582c7a74454d5cb654a8dfaf605b81393/skimage/segmentation/morphsnakes.py#L155): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 317:32](https://github.com/scikit-image/scikit-image/blob/4adaadc582c7a74454d5cb654a8dfaf605b81393/skimage/segmentation/morphsnakes.py#L317): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 317:58](https://github.com/scikit-image/scikit-image/blob/4adaadc582c7a74454d5cb654a8dfaf605b81393/skimage/segmentation/morphsnakes.py#L317): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

stefanv commented:
I've used your code before and have achieved excellent results, so I am delighted to see your PR, @pmneila!

emmanuelle commented on doc/examples/segmentation/plot_morphsnakes.py:
```diff
@@ -0,0 +1,120 @@
+**MorphGAC** is suitable for images with visible contours, even when these
+contours might be noisy, cluttered, or partially unclear. It requires, however,
+that the image is preprocessed to highlight the contours. This can be done
+using the function ``gborders``, although the user might want to define its own
```
define their own version?

pmneila replied:
Right.

emmanuelle commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,451 @@
+    return np.array(dilations, dtype=np.int8).min(0)
+
+
+def SIoIS(u):
```
Would it be possible to have a more meaningful name for this function? I guess it's an acronym, so maybe the full name instead?

pmneila replied:
This one is tricky. 'o' is the composition operator. SIoIS is the composition of IS followed by SI, and "SIoIS" is the name we gave to that operator in the paper. I can call it sup_inf_inf_sup (and inf_sup_sup_inf for ISoSI), but I'm not sure if it will be clearer that way…

pmneila replied:
Alternatively, since those functions are never called explicitly, I can remove them and rewrite `curvop` as

```python
curvop = _fcycle([lambda u: sup_inf(inf_sup(u)),   # SIoIS
                  lambda u: inf_sup(sup_inf(u))])  # ISoSI
```

emmanuelle commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,451 @@
+_P3[8][[0, 1, 2], [2, 1, 0], :] = 1
+
+
+def SI(u):
```
sup_inf instead of SI?

pmneila replied:
Fine.

pmneila commented:
@stefanv Thank you!

pmneila commented:
I made the changes proposed by @emmanuelle.

Also, I removed `SIoIS` and `ISoSI` and rewrote `curvop` using lambda functions. Not sure if this approach is better.

stefanv commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+           2014, DOI 10.1109/TPAMI.2013.106
+    """
+
+    # TODO: Make it work with "color" (or multi-channel) images.
```
Is this still on the horizon?

jni replied:
In general, we don't have TODOs in our code (is this true? I think it's true). Instead, remove this line, and raise an issue on GitHub describing the feature request, once this is merged.

pmneila replied:
It is on the horizon, but not in the short term. I'll remove the TODO line.

stefanv commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+    return res
+
+
+def gborders(image, alpha=100.0, sigma=5.0):
```
This function name is somewhat unintuitive---do you have any other suggestions?

Also, the one line description "Stopping criterion for image borders." does not make its function immediately clear to me.

jni replied:
I agree. How about `inverse_gaussian_gradient`?

pmneila replied:
I like `inverse_gaussian_gradient`. I'll use that.

jni commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+        radius = min(image_shape) * 3.0 / 8.0
+
+    grid = np.mgrid[[slice(i) for i in image_shape]]
+    grid = (grid.T - center).T
```
Is this correct? `grid.T - center` will subtract the row center from the column coordinates and the column center from the row coordinates?

An alternative:

```python
grid -= np.reshape(center, (ndim,) + (1,) * ndim)
```

or:

```python
grid = np.mgrid[[slice(-center[i], image_shape[i] - center[i]) for i in range(ndim)]]
```

pmneila replied:
`grid.T - center` is tricky, but it is surprisingly correct (I had to check it).

Note that `grid[:. i, j]` contains the coordinate in format (row, column). Also, `grid[:, i, j] == grid.T[j, i, :]` for all `i`, `j`. That means that `grid.T[j, i, :]` also contains the coordinate in format (row, column). For example,

```python
In [25]: grid = np.mgrid[:3, :4]

In [26]: grid[:, 2, 3]
array([2, 3])

In [27]: grid.T[3, 2, :]
array([2, 3])
```

Therefore, subtracting `center` from `grid.T` is correct.

In any case, if you prefer the `reshape` approach for being less confusing, I'm open to change it.

jni replied:
Oh of course! I confused transposing with flipping. Personally I find the `-=` approach to be the most readable, but change only if you agree... I can see how the reshape argument is itself confusing.

jni replied:
Thanks for the long exposition. =)

jni commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+    """
+
+    grid = np.mgrid[[slice(i) for i in image_shape]]
+    grid = (grid // square_size) & 1
```
What does `& 1` do here?

jni replied:
If it's converting to bool, please use `.astype(bool)`

pmneila replied:
`& 1` is the way to make the square pattern of the checkerboard (it is like writing `% 2`). For example, in 1D, if `square_size` is 3, `grid // square_size` would be something like

`[0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3...]`

and `(grid // square_size) & 1` would be

`[0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1...]`

which is a one-dimensional checkerboard. It should be integer, not `bool`.

jni replied:
OooooOOOOOhhhh! Of course! Maybe change it to `% 2` and/or add a comment, "alternate 0/1 for even/odd numbers"?

pmneila replied:
Following your suggestion, I added the comment and kept `& 1` (since `& 1` is several times faster than `% 2`). 

```python
In [6]: aux = np.random.randint(100, size=(1000000))

In [7]: %timeit aux & 1
1000 loops, best of 3: 907 µs per loop

In [8]: %timeit aux % 2
100 loops, best of 3: 18.1 ms per loop

In [9]: np.all((aux & 1) == (aux % 2))
Out[9]: True
```

jni commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+
+    Parameters
+    ----------
+    gimage : (M, N) array
```
From my reading of the code, this could actually be 3D, right?

pmneila commented on doc/examples/segmentation/plot_morphsnakes.py:
```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from skimage import data, img_as_float
+from skimage.segmentation import (morph_acwe, morph_gac, gborders,
```
I'm fine with changing the names to remove acronyms. However, I don't like the option of having a single function with a `mode` argument, since `morph_acwe` and `morph_gac` do not share the same parameters (for example, `lambda1` and `lambda2` for `morph_acwe` vs. `threshold` and `balloon` for `morph_acwe`).

`chan_vese` is the same than `active_contours_without_edges`, and there is already an implementation of the classical Chan-Vese called `chan_vese` in `scikit-image`. So, what about `morphological_chan_vese` and `morphological_geodesic_active_contour`?

jni replied:
I like these suggestions a lot, thanks!

pmneila commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+
+
+curvop = _fcycle([lambda u: sup_inf(inf_sup(u)),   # SIoIS
+                  lambda u: inf_sup(sup_inf(u))])  # ISoSI
```
It doesn't need to be a public function. I'll change the name to `_curvop`.

pmneila commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+def _check_input(image, init_level_set):
+    """Check that shapes of `image` and `init_level_set` match."""
+    if len(image.shape) not in [2, 3]:
+        raise ValueError("Input image is not a 2D or 3D array.")
```
Ok. I'll change that.

pmneila commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,440 @@
+
+def _find_threshold(image, percentile=20):
+    """Find a proper threshold for `morph_gac` using the given percentile."""
+    return np.percentile(image, percentile)
```
It is. Initially it was a little bit more complex but after some simplifications it got like that. I'll remove it.

pmneila commented:
Regarding the comments about 3D volumes of @stefanv and @jni, I thought that `scikit-learn` did not support them. That's why the documentation and the testing code does not include any reference to 3D volumes (even when the code should work with them).

If 3D volumes are ok, is there any operator or any official way in `scikit-learn` for distinguishing a 3D volume from a multichannel 2D image?

jni commented:
@pmneila I assume you mean scikit-image. =D

My [self-appointed title is Guardian of the Dimensions](https://github.com/scikit-image/scikit-image/pull/1038#issuecomment-47189573), because in fact I *always* try to push PRs to support 3D and even nD where possible. (@emmanuelle is Vice-Guardian. =P) There's many, many functions supporting nD in scikit-image (see the [release notes](http://scikit-image.org/docs/dev/release_notes_and_installation.html), for example, or grep for "3D" in the repo). The main reason that there are not more examples in the docs is that 3D viz in Python is sadly lacking, so it's hard to demonstrate these tools. If you have suggestions about this, we would welcome them!

Finally, yes, the convention is `multichannel={True,False}`. If `True`, the final dimension is considered channels. Again, feel free to grep for this for some examples. =)

pmneila commented:
@jni Yes, of course, I meant `scikit-image` haha

I added a new test with a 3D volume and updated the documentation to reflect the fact that both images and volumes are supported. I changed the names of the functions to make them clearer and with no acronyms. I also followed the rest of suggestions (I think I didn't forget any). 

Still no multichannel/color support in `morphological_chan_vese`. That will come at some point in the future.

stefanv commented:
Thank you, @pmneila---this is an exciting new addition!

pmneila commented:
Thank you all for your suggestions!

soupault commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,442 @@
+                         "match the dimensions of the image.")
+
+
+def _init_level_set(init_level_set, image_shape):
```
`level_set`s are partially duplicated by those in `chan_vese`. We should consider factoring them out.

soupault commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,442 @@
+    """Check that shapes of `image` and `init_level_set` match."""
+    assert_nD(image, [2, 3])
+
+    if len(image.shape) != len(init_level_set.shape):
```
`len(something.shape)` is just `something.ndim`

pmneila replied:
You are right.

soupault commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,442 @@
+
+    Parameters
+    ----------
+    gimage : (M, N) or (L, M, N) array
```
Our naming convention says this should be `image`

pmneila replied:
This is on purpose. This argument is rarely the image the user wants to work with, but a transformation of that image as explained in the docstring of the function. I called it `gimage` instead of `image` to hint the user about reading the docstring. It can be changed if the naming convention is critical, though.

soupault commented on skimage/segmentation/morphsnakes.py:
```diff
@@ -0,0 +1,442 @@
+        preprocessing.
+    iterations : uint
+        Number of iterations to run.
+    init_level_set : str, (M, N) array, or (L, M, N) array
```
`, optional`

pmneila replied:
Yes.

stefanv commented:
@soupault For minor additions like this that come after a PR is more-or-less concluded, feel free to simply push the changes into the contributor branch.  Otherwise, we risk having entire PRs stall for a small benefit.

stefanv commented:
@soupault To be fair, it wasn't clear that the review was almost concluded, because @jni's approval appeared way earlier in the comment stream.

I am building a small command line tool with some students to make it easier to pull and push to PRs.
